Gurobipy Environment Handling

torressa opened the pull request:
- Fix coin-or/pulp/issues/586.
- For coin-or/pulp/issues/571
  - Fix issue with `pulp.list_solvers`. Cheers @simonbowly!
  - Added three new parameters to `GUROBI` that allow users more fine control over the environments used.
    - `env: gp.Env` users environment to use.
    - `envOptions: dict`, options to give to the environment if `manageEnv` is True.
    - `manageEnv: bool`: whether environments are handled internally or not default is false (as it was before).
   - Added a new function to free the environments and models properly `close`. If users are using their own environments or using `manageEnv=True`, they should always call this function afterwards.


Sample use: 
```python
# Pass environment as a parameter
with gp.Env(params=env_options) as env:
  prob = generate_lp()
  solver = GUROBI(msg=True, env=env, **solver_options)
  prob.solve(solver)
  # Dispose model (required to correctly free env)
  solver.close()
```

Or, 
```python
# Use manageEnv=True and pulp will take care
solver = GUROBI(msg=True, manageEnv=True, envOptions=env_options, **solver_options)
prob = generate_lp()
prob.solve(solver)
# Dispose env + model
solver.close()
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/coin-or/pulp?pullRequest=588) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/coin-or/pulp?pullRequest=588) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: torressa<br/>:white_check_mark: djunglas<br/>:x: simonbowly<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/coin-or/pulp?pullRequest=588) it.</sub>

EwoutH commented:
This looks quite cool! What's needed to move this further? A review from @pchtsp?

Tomkourou commented:
Commenting to move this along. Problems with the gurobipy single-use licenses are still unresolved.

torressa commented on pulp/apis/gurobi_api.py:
```diff
@@ -131,46 +158,66 @@ def findSolutionValues(self, lp):
-                ):
-                    constr.pi = value
-
+            if status in [constants.LpStatusOptimal, constants.LpStatusNotSolved]:
```
Addressed in b95e06143c7d20ac817bfc21bf31730ed1cb450d

torressa commented:
@simon-b I've just added some usage explanations to the doc-string, should be clearer now.
(tested building the docs on my machine and was OK). 
Also rebased master, hope this is merged soon!

torressa commented on pulp/apis/gurobi_api.py:
```diff
@@ -197,20 +293,24 @@ def buildSolverModel(self, lp):
-                    lowBound, upBound, vtype=varType, obj=obj, name=var.name
-                )
+                    varType = gp.GRB.INTEGER
+                # only add variable once, ow new variable will be created.
```
Added this check for https://github.com/coin-or/pulp/blob/a7684ac90d93dc3bca382d90841a9ab5f398b50f/pulp/tests/test_pulp.py#L660
Otherwise `x` is added twice and the test fails.

simon-b reviewed (approved):
Looks great to me! Test on my side with various options for managing the gurobi env and all seem to work properly.

Suggest a review from @pchtsp and hopefully can get this closed soon !

torressa commented on pulp/tests/test_pulp.py:
```diff
@@ -1250,9 +1254,15 @@ def test_invalid_var_names(self):
-                prob, self.solver, [const.LpStatusOptimal], {x: 4, y: -1, z: 6, w: 0}
-            )
+            if self.solver.name not in [
+                "GUROBI_CMD",  # end is a key-word for LP files
```
I think this must've not been tested using `GUROBI_CMD`. Variable `z` is called `End` which is a keyword for LP files hence results in the LP file being solved being pretty different 

```
Minimize
  9
Subject To
Bounds
End
```

(I versions of Gurobi back to 8.0.0 with the same result)

torressa commented:
The tests should now run using the Gurobi license included when installing `gurobipy` via pip install (which I've included in the workflow). There is one test where the size limitation is reached, for which I've added a decorator. This test will run fine if a full license is picked up.
Unfortunately, the pip license will not be picked up by `gurobi_cl` or the `GUROBI_CMD` interface.

stumitchell commented:
@torressa we are getting a test fail now,
Also can we remove the restricted licence warning from the test output?

torressa commented:
Hi @stumitchell! If you mean the tests on the CI, I am not sure why, the errors are:
```
 File "/opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/site-packages/pulp/apis/highs_api.py", line 327, in buildSolverModel
    lp.solverModel.addCol(
ModuleNotFoundError: No module named 'numpy'
```

This is weird cause it passes on macos and also cause highspy doesn't use numpy

stumitchell commented:
Right, I see the issue I have contacted @aphi about it

We probably want to get master fixed before we merge anything new

pchtsp commented:
Thanks @stumitchell 

I was waiting (too) patiently for the numpy issue to resolve in highspy. I did forget to create an issue there :  ) so thanks for doing so.

I've currently taken out the installation of the highspy python package from the github CI on the master branch. This should solve this issue.

@torressa can you re-merge master please? Sorry for the lack of response, these months have been very busy.

F.

aphi commented:
Looks like we both investigated this at the same time 😬.

I have a PR which keeps the testcase for highspy but just includes numpy, which also resolves the issue.

https://github.com/coin-or/pulp/pull/677

torressa commented:
No worries @pchtsp! Merged.
I just opened an issue in HiGHS repo so that they are aware: https://github.com/ERGO-Code/HiGHS/issues/1378
